File: Semana 2/Jueves 4-09/oop-sample/src/main/java/acme/oop/shared/domain/model/valueobjects/Address.java
package acme.oop.shared.domain.model.valueobjects;

import java.util.Objects;

/**
 * Represents a physical address with street, city, postal code, and country.
 * All fields are mandatory and cannot be null or blank.
 * @param street the street address, it must not be null or blank.
 * @param city the city name, it must not be null or blank.
 * @param postalCode the postal code, it must not be null or blank.
 * @param country the country name, it must not be null or blank.
 * @author Open Source Application Development Team
 */
public record Address(String street, String city, String postalCode, String country) {
    /**
     * Constructs an Address object with validation.
     * @param street the street address, it must not be null or blank.
     * @param city the city name, it must not be null or blank.
     * @param postalCode the postal code, it must not be null or blank.
     * @param country the country name, it must not be null or blank.
     * @throws IllegalArgumentException if any of the parameters is null or blank.
     */
    public Address {
        if (street == null || street.isBlank()) {
            throw new IllegalArgumentException("Street cannot be null or blank");
        }
        if (Objects.isNull(city) || city.isBlank()) {
            throw new IllegalArgumentException("City cannot be null or blank");
        }
        if (Objects.isNull(postalCode)  || postalCode.isBlank()) {
            throw new IllegalArgumentException("Postal code cannot be null or blank");
        }
        if (Objects.isNull(country)  || country.isBlank()) {
            throw new IllegalArgumentException("Country cannot be null or blank");
        }
    }

    public String toString() {
        return street + ", " + city + ", " + postalCode + ", " + country;
        //return String.format("%s, %s, %s, %s", street, city, postalCode, country);
    }
}

File: Semana 2/Sabado 6-09/oopsample/ACME.OOP/Shared/Domain/Model/ValueObjects/Address.cs
namespace ACME.OOP.Shared.Domain.Model.ValueObjects;

// <summary>
// Representa una direccion internacional (Address)
// </summary>

public record Address
{
    public string Street { get; init; }
    public string Number { get; init; }
    public string City { get; init; }
    public string? StateorRegion { get; init; }
    public string PostalCode { get; init; }
    public string Country { get; init; }
    
    /// <summary>
    /// Creamos una instancia de address
    /// </summary>
    /// <param name = "street"> direccion de calle 
    

    public Address(string street, string number, string city, string? stateorRegion, string postalCode, string country)
    {
        if (string.IsNullOrWhiteSpace(street))
            throw new ArgumentException("Street cannot be null or empty.", nameof(street));
        if (string.IsNullOrWhiteSpace(number))
            throw new ArgumentException("Number cannot be null or empty.", nameof(number));
        if (string.IsNullOrWhiteSpace(city))
            throw new ArgumentException("City cannot be null or empty.", nameof(city));
        if(string.IsNullOrEmpty(postalCode))
            throw new ArgumentException("Postal code cannot be null or empty.", nameof(postalCode));
        if (string.IsNullOrEmpty(country))
            throw new ArgumentException("Country cannot be null or empty.", nameof(country));
        
        Street = street;
        Number = number;
        City = city;
        StateorRegion = stateorRegion;
        PostalCode = postalCode;
        Country = country;
    }
    
    public override string ToString()
     => $"{Street} {Number}, {City}, {(string.IsNullOrWhiteSpace(StateorRegion) ? "" : StateorRegion + ", ")}{PostalCode}, {Country}";
};